lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp

```diff
@@ -0,0 +1,284 @@
+    inlineAnnoClass =
+        StringAttr::get(cx, "firrtl.transforms.BlackBoxInlineAnno");
+    pathAnnoClass = StringAttr::get(cx, "firrtl.transforms.BlackBoxPathAnno");
+    resourceAnnoClass =
+        StringAttr::get(cx, "firrtl.transforms.BlackBoxResourceAnno");
+
+    // Run on the CircuitOp in the module.
+    for (auto &op : *getOperation().getBody())
+      if (auto circuitOp = dyn_cast<CircuitOp>(&op))
+        runOnCircuit(circuitOp);
+  }
+
+  void runOnCircuit(CircuitOp circuitOp) {
+    // Determine the target directory and resource file name from the
+    // annotations present on the circuit operation.
+    StringRef specifiedTargetDir = ".";
+    StringRef specifiedResourceFileName = "firrtl_black_box_resource_files.f";
+
+    if (auto attrs = getAnnotations(circuitOp)) {
+      for (auto attr : attrs) {
+        auto anno = attr.dyn_cast<DictionaryAttr>();
+        if (!anno)
+          continue;
+        auto cls = anno.getAs<StringAttr>("class");
+
+        // Handle target dir annotation.
+        if (cls == targetDirAnnoClass) {
+          auto x = anno.getAs<StringAttr>("targetDir");
+          if (!x) {
+            circuitOp->emitError(targetDirAnnoClass.getValue() +
+                                 " annotation missing \"targetDir\" attribute");
+            signalPassFailure();
+            continue;
+          }
+          specifiedTargetDir = x.getValue();
+          continue;
+        }
+
+        // Handle resource file name annotation.
+        if (cls == resourceFileNameAnnoClass) {
+          auto x = anno.getAs<StringAttr>("resourceFileName");
+          if (!x) {
+            circuitOp->emitError(
+                resourceFileNameAnnoClass.getValue() +
+                " annotation missing \"resourceFileName\" attribute");
+            signalPassFailure();
+            continue;
+          }
+          specifiedResourceFileName = x.getValue();
+          continue;
+        }
+      }
+    }
+
+    targetDir = "";
+    appendPossiblyAbsolutePath(targetDir, specifiedTargetDir);
+    resourceFileName = "";
+    appendPossiblyAbsolutePath(resourceFileName, specifiedResourceFileName);
+
+    LLVM_DEBUG(llvm::dbgs()
+               << "Black box target directory: " << targetDir << "\n"
+               << "Black box resource file name: " << resourceFileName << "\n");
+
+    // Create the target directory if it does not exist.
+    if (auto error = llvm::sys::fs::create_directory(targetDir)) {
+      circuitOp->emitError("cannot create black box target directory \"" +
+                           targetDir + "\": " + error.message());
+      return signalPassFailure();
+    }
+
+    // Gather the relevant annotations on all modules in the circuit so we can
+    // process them in parallel afterwards.
+    OpBuilder builder(&getContext());
+    builder.setInsertionPointToEnd(getOperation().getBody());
+
+    for (auto &op : *circuitOp.getBody()) {
+      if (!isa<FModuleOp>(op) && !isa<FExtModuleOp>(op))
```

Comment:
The Scala implementation says ["Handle source for Verilog ExtModules (BlackBoxes)"](https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala#L52), but the actual implementation just gathers the annotations which can [target anything with a `ModuleName`](https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala#L20). I figure in practice people will only put this on `ExtModule`s, but this prevents any surprises where SFC accepts something which MFC rejects. Especially since we actually don't really rely on the operation the annotation sits on.